template-coq/src/template_monad.ml

```diff
@@ -5,408 +5,366 @@ open Pp
-      | _ -> monad_failure "tmCurrentModPath" 1
-
-    else if Globnames.eq_gr glob_ref ptmMkDefinition then
-      match args with
-      | name::body::[] ->
-        (TmMkDefinition (name, body), universes)
-      | _ -> monad_failure "tmMkDefinition" 2
-
-    else if Globnames.eq_gr glob_ref ptmQuote then
-      match args with
-      | _::trm::[] ->
-        (TmQuote (false,trm), universes)
-      | _ -> monad_failure "tmQuote" 2
-    else if Globnames.eq_gr glob_ref ptmQuoteRec then
-      match args with
-      | _::trm::[] ->
-        (TmQuote (true,trm), universes)
-      | _ -> monad_failure "tmQuoteRec" 2
-
-    else if Globnames.eq_gr glob_ref ptmQuoteInductive || Globnames.eq_gr glob_ref ttmQuoteInductive then
-      match args with
-      | name::[] ->
-        (TmQuoteInd name, universes)
-      | _ -> monad_failure "tmQuoteInductive" 1
-    else if Globnames.eq_gr glob_ref ptmQuoteUniverses || Globnames.eq_gr glob_ref ttmQuoteUniverses then
-      match args with
-      | [] ->
-        (TmQuoteUnivs, universes)
-      | _ -> monad_failure "tmQuoteUniverses" 0
-    else if Globnames.eq_gr glob_ref ptmQuoteConstant || Globnames.eq_gr glob_ref ttmQuoteConstant then
-      match args with
-      | name::bypass::[] ->
-        (TmQuoteConst (name, bypass), universes)
-      | _ -> monad_failure "tmQuoteConstant" 2
-
-    else if Globnames.eq_gr glob_ref ptmMkInductive then
-      match args with
-      | mind::[] -> (TmMkInductive mind, universes)
-      | _ -> monad_failure "tmMkInductive" 1
-    else if Globnames.eq_gr glob_ref ttmInductive then
-      match args with
-      | mind::[] -> (TmMkInductive mind, universes)
-      | _ -> monad_failure "tmInductive" 1
-    else if Globnames.eq_gr glob_ref ptmUnquote then
-      match args with
-      | t::[] ->
-        (TmUnquote t, universes)
-      | _ -> monad_failure "tmUnquote" 1
-    else if Globnames.eq_gr glob_ref ptmUnquoteTyped then
-      match args with
-      | typ::t::[] ->
-        (TmUnquoteTyped (typ, t), universes)
-      | _ -> monad_failure "tmUnquoteTyped" 2
-
-    else if Globnames.eq_gr glob_ref ptmExistingInstance || Globnames.eq_gr glob_ref ttmExistingInstance then
-      match args with
-      | name :: [] ->
-        (TmExistingInstance name, universes)
-      | _ -> monad_failure "tmExistingInstance" 1
-    else if Globnames.eq_gr glob_ref ptmInferInstance then
-      match args with
-      | s :: typ :: [] ->
-        (TmInferInstance (s, typ), universes)
-      | _ -> monad_failure "tmInferInstance" 2
-    else if Globnames.eq_gr glob_ref ttmInferInstance then
-      match args with
-      | typ :: [] ->
-        (TmInferInstanceTerm typ, universes)
-      | _ -> monad_failure "tmInferInstance" 2
-
-    else CErrors.user_err (str "Invalid argument or not yet implemented. The argument must be a TemplateProgram: " ++ pr_constr coConstr)
-
-end
+     ptmUnquote,
+     ptmUnquoteTyped,
+
+     ptmInferInstance,
+     ptmExistingInstance,
+
+     ptmTestQuote,
```

Comment:
It seems like you "broke the abstraction" simply to add these 4 definitions. Why not just put them in another file? Unless I'm missing something, they don't seem to be used anywhere else in the file.

Reply:
Or I can add an .mli file. I will try this.

Reply:
I like the `.mli` idea. Very little of the file actually needs to be exposed.